pkg/kvstore/store/syncstore_test.go

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package store
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"k8s.io/client-go/util/workqueue"
+)
+
+type KVPair struct{ Key, Value string }
+
+func NewKVPair(key, value string) *KVPair      { return &KVPair{Key: key, Value: value} }
+func (kv *KVPair) GetKeyName() string          { return kv.Key }
+func (kv *KVPair) Marshal() ([]byte, error)    { return []byte(kv.Value), nil }
+func (kv *KVPair) Unmarshal(data []byte) error { return nil /* not used */ }
+
+type fakeBackend struct {
+	t *testing.T
+
+	updated chan *KVPair
+	deleted chan *KVPair
+
+	errorsOnUpdate map[string]uint
+	errorsOnDelete map[string]uint
+}
+
+func NewFakeBackend(t *testing.T) *fakeBackend {
+	return &fakeBackend{
+		t: t,
+
+		updated: make(chan *KVPair),
+		deleted: make(chan *KVPair),
+
+		errorsOnUpdate: make(map[string]uint),
+		errorsOnDelete: make(map[string]uint),
+	}
+}
+
+func (fb *fakeBackend) Update(ctx context.Context, key string, value []byte, _ bool) error {
+	select {
+	case fb.updated <- NewKVPair(key, string(value)):
+	case <-ctx.Done():
+		require.Failf(fb.t, "Context closed before writing to updated channel", "key: %s, value: %s", key, value)
+	}
+
+	if cnt := fb.errorsOnUpdate[key]; cnt > 0 {
+		fb.errorsOnUpdate[key]--
+		return errors.New("failing on purpose")
+	}
+
+	return nil
+}
+
+func (fb *fakeBackend) Delete(ctx context.Context, key string) error {
+	select {
+	case fb.deleted <- NewKVPair(key, ""):
+	case <-ctx.Done():
+		require.Failf(fb.t, "Context closed before writing to deleted channel", "key: %s", key)
+	}
+
+	if cnt := fb.errorsOnDelete[key]; cnt > 0 {
+		fb.errorsOnDelete[key]--
+		return errors.New("failing on purpose")
+	}
+
+	return nil
+}
+
+func eventually(in <-chan *KVPair, timeout time.Duration) *KVPair {
+	select {
+	case kv := <-in:
+		return kv
+	case <-time.After(timeout):
+		return NewKVPair("error", "timed out waiting for KV")
+	}
+}
+
+func TestNewWorkqueueSyncStore(t *testing.T) {
+	backend := NewFakeBackend(t)
+
+	store := NewWorkqueueSyncStore(backend, "/foo/bar").(*wqSyncStore)
+	require.Equal(t, backend, store.backend)
+	require.Equal(t, "/foo/bar", store.prefix)
+	require.Equal(t, workqueue.DefaultControllerRateLimiter(), store.limiter)
+	require.Equal(t, uint(1), store.workers)
+	require.Equal(t, true, store.withLease)
+
+	limiter := workqueue.DefaultItemBasedRateLimiter()
+	store = NewWorkqueueSyncStore(backend, "/foo/bar", WSSWithRateLimiter(limiter), WSSWithWorkers(10), WSSWithoutLease()).(*wqSyncStore)
+	require.Equal(t, limiter, store.limiter)
+	require.Equal(t, uint(10), store.workers)
+	require.Equal(t, false, store.withLease)
```

Comment:
I would prefer to test behavior rather than checking particular fields - this type of tests makes it harder to change names, refactor code etc and any refactor can actually break behavior without test failing - for example, let's say that  `withLease` is not used by SyncStore implementation at all and test will be still passing. 
To test withLease, we could have the assertion that update to backend is done with lease or without, depending on the option. 
For workers, we could have 2 workers and make operation to backend blocking and see if we have two concurrent operations. And so on.